common/predictive-text/worker/word_breaking/default-word-breaking/index.ts

```diff
@@ -0,0 +1,369 @@
+// Include the word-breaking data here:
+/// <reference path="./data.ts" />
+namespace wordBreakers {
+  /**
+   * Word breaker based on Unicode Standard Annex #29, Section 4.1:
+   * Default Word Boundary Specification.   
+   * 
+   * @see http://unicode.org/reports/tr29/#Word_Boundaries
+   * @see https://github.com/eddieantonio/unicode-default-word-boundary/tree/v12.0.0
+   */
+  export function default_(text: string): Span[] {
+    let spans = Array.from(findSpans(text));
+    return spans.filter(span => isNonSpace(span.text));
+  }
+
+  // Utilities //
+
+  /**
+   * Generator that yields every successive span from the the text.
+   * @param text Any valid USVString to segment.
+   */
+  function* findSpans(text: string): Iterable<Span> {
+    // TODO: don't throw the boundaries into an array.
+    let boundaries = Array.from(findBoundaries(text));
+    if (boundaries.length == 0) {
+      return;
+    }
+
+    // All non-empty strings have at least TWO boundaries at the start and end of
+    // the string.
+    for (let i = 0; i < boundaries.length - 1; i++) {
+      let start = boundaries[i];
+      let end = boundaries[i + 1];
+      yield new LazySpan(text, start, end);
+    }
+  }
+
+  /**
+   * A span that does not cut out the substring until it absolutely has to!
+   */
+  class LazySpan implements Span {
+    private _source: string;
+    readonly start: number;
+    readonly end: number;
+    constructor(source: string, start: number, end: number) {
+      this._source = source;
+      this.start = start;
+      this.end = end;
+    }
+
+    get text(): string {
+      return this._source.substring(this.start, this.end);
+    }
+
+    get length(): number {
+      return this.end - this.start;
+    }
+  }
+
+  /**
+   * Returns true when the chunk does not solely consist of whitespace.
+   *
+   * @param chunk a chunk of text. Starts and ends at word boundaries.
+   */
+  function isNonSpace(chunk: string): boolean {
+    return !Array.from(chunk).map(property).every(wb => (
+      wb === 'CR' || wb === 'LF' || wb === 'Newline' || wb === 'WSegSpace'
+    ));
+  }
+
+  /**
+   * Yields a series of string indices where a word break should
+   * occur. That is, there should be a break BEFORE each string
+   * index yielded by this generator.
+   *
+   * @param text Text to find word boundaries in.
+   */
+  function* findBoundaries(text: string): Iterable<number> {
```

Comment:
`function*`?

Reply:
It's a [generator function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function*)

Instead of doing the conventional `var a = []; a.push(x)` this _yields_ values one by one. In this case, it yields the string indices before which is a word boundary. In my original implementation, I intended this to be a generator to get indices one by one as required, but since I use `Array.from()` to yield all values and map over that array, using a generator is not useful now.

I don't think this is the IE11 issue, because I was looking at TS's compile sources, and it actually removes the `function*` syntax, replacing it with a polyfill. e.g., take a look at `embedded_worker.js`.

Reply:
Ah, okay.  I tried looking it up when I saw it, but trying to Google search for `function*` doesn't exactly give that result; thanks!